src/dataframe/dataframe.jl

```diff
@@ -104,37 +105,44 @@ struct DataFrame <: AbstractDataFrame
             throw(DimensionMismatch("Number of columns ($(length(columns))) and number of" *
                                     " column names ($(length(colindex))) are not equal"))
         end
-        lengths = [isa(col, AbstractArray) ? length(col) : 1 for col in columns]
-        minlen, maxlen = extrema(lengths)
-        if minlen != maxlen || minlen == maxlen == 1
-            # recycle scalars
-            for (i, col) in enumerate(columns)
-                if col isa Union{AbstractArray{<:Any, 0}, Ref}
-                    x = col[]
-                    columns[i] = fill!(Tables.allocatecolumn(typeof(x), maxlen), x)
-                    lengths[i] = maxlen
-                elseif !(col isa AbstractArray)
-                    columns[i] = fill!(Tables.allocatecolumn(typeof(col), maxlen), col)
-                    lengths[i] = maxlen
+
+        len = -1
+        firstvec = -1
+        for (i, col) in enumerate(columns)
+            if col isa AbstractVector
+                if len == -1
+                    len = length(col)
+                    firstvec = i
+                elseif len != length(col)
+                    n1 = _names(colindex)[firstvec]
+                    n2 = _names(colindex)[i]
+                    throw(DimensionMismatch("column :$n1 has length $len and column " *
+                                            ":$n2 has length $(length(col))"))
                 end
             end
-            uls = unique(lengths)
-            if length(uls) != 1
-                strnames = string.(names(colindex))
-                estrings = ["column length $u for column(s) " *
-                            join(strnames[lengths .== u], ", ", " and ") for (i, u) in enumerate(uls)]
-                throw(DimensionMismatch(join(estrings, " is incompatible with ", ", and is incompatible with ")))
-            end
         end
-        for (i, c) in enumerate(columns)
-            if isa(c, AbstractRange)
-                columns[i] = collect(c)
-            elseif !isa(c, AbstractVector)
-                throw(DimensionMismatch("columns must be 1-dimensional"))
-            elseif copycols
-                columns[i] = copy(c)
+        len == -1 && (len = 1) # we got no vectors so make one row of scalars
+
+        # we write into columns as we know that it is guarantted
```

Comment:
It doesn't sound right for an inner to constructor to rely on the behavior of outer constructors, since it can be called directly by user code. Is there another solution?

```suggestion
        # we write into columns as we know that it is guaranteed
```

Reply:
> since it can be called directly by user code

it cannot, as we do not expose `Index`.

> It doesn't sound right for an inner to constructor to rely on the behavior of outer constructors

It already did - I simply corrected a bug in an outer constructor.

> Is there another solution?

Yes, but it would involve making an additional copy of `columns` vector inside an inner constructor, which I understand we wanted to avoid (outer constructors now already alsways make copies and these copies cannot be avoided there).

Reply:
OK, let's say it's an internal outer constructor then. :-D